### PR TITLE
add `minVersion` to requirement

### PR DESCRIPTION
**Pull request by onigoetz**
Hi,

as mentioned in #1 this PR adds a new method to `Requirement` that allows to request for the minimum version that fullfills a requirement.

All tests from the nodejs version were added to the test suite

Three changes were needed to get it to pass correctly:
1) The addition of the `minVersion` method to `Requirement`
2) The addition of a `nextIncrement` method to `Semver` that allows to add the smallest increment possible to a version
   - if a version is simply 1.0.0 it will happily increment to 1.0.1
   - if a version is 1.0 it will increment to 1.1
   - if a version is 1 it will increment to 2
   - suffixes are also taken in account might need some improvements though
 3) The tokenizer treated "-" as a hyphen (as in version ranges) in all cases, though the NPM ranges spec mentions that a hyphen must have a space before and after to be valid, the change adds this behaviour to the tokenizer

Best regards,
Stéphane

### CODE REVIEW

**Comment by onigoetz** [comment]
Hi, @Legion2 @erksch,
Sorry to bother you directly, I'm just gently nudging if you have seen my PR ?

Thanks in advance

File: src/commonMain/kotlin/de/voize/semver4k/Requirement.kt [hunk 1]
```diff
@@ -87,6 +87,61 @@ class Requirement
...
+        }
+
+        if (minver != null && this.isSatisfiedBy(minver)) {
+            return minver
+        }
+
+        return null
+    }
+
+    private fun getAllAllRanges(requirement: Requirement?, res: MutableList<Range?>): List<Range?> {
```
Comment (Legion2) [inline src/commonMain/kotlin/de/voize/semver4k/Requirement.kt]: maybe give it another name

File: src/commonMain/kotlin/de/voize/semver4k/Semver.kt [hunk 2]
```diff
@@ -470,6 +470,23 @@ class Semver @JvmOverloads constructor(val originalValue: String, val type: Semv
...
+    fun nextIncrement(): Semver {
+        if (suffixTokens == null || suffixTokens.isEmpty()) {
+            if (patch != null) {
+                return withIncPatch()
+            }
+
+            if (minor != null) {
+                return create(type, major!!, minor!!.plus(1), null, null, build);
+            }
+                return create(type, major!!.plus(1), null, null, null, build);
```
Comment (Legion2) [inline src/commonMain/kotlin/de/voize/semver4k/Semver.kt]: code must be formatted

File: src/commonMain/kotlin/de/voize/semver4k/Semver.kt [hunk 3]
```diff
@@ -470,6 +470,23 @@ class Semver @JvmOverloads constructor(val originalValue: String, val type: Semv
...
+
+        }
+
+        val suffixes = suffixTokens.joinToString(separator = ".")
+        return withSuffix(suffixes + ".0")
+    }
```
Comment (Legion2) [inline src/commonMain/kotlin/de/voize/semver4k/Semver.kt:495]: can you add test cases for nextIncrement

**Comment by onigoetz** [comment]
I just saw that my change to the tokenizer broke the npm tokenizer tests, it's fixed now

**Review by Legion2 — CHANGES_REQUESTED** [review]

File: src/commonMain/kotlin/de/voize/semver4k/Requirement.kt (hunk 1 above)
Comment (onigoetz) [reply src/commonMain/kotlin/de/voize/semver4k/Requirement.kt]: oops, right, I forgot to rename that after experimenting, I will do that

File: src/commonMain/kotlin/de/voize/semver4k/Semver.kt (hunk 3 above)
Comment (onigoetz) [reply src/commonMain/kotlin/de/voize/semver4k/Semver.kt:495]: tests added

File: src/commonMain/kotlin/de/voize/semver4k/Semver.kt (hunk 2 above)
Comment (onigoetz) [reply src/commonMain/kotlin/de/voize/semver4k/Semver.kt]: done

File: src/commonMain/kotlin/de/voize/semver4k/Requirement.kt (hunk 1 above)
Comment (onigoetz) [reply src/commonMain/kotlin/de/voize/semver4k/Requirement.kt]: done, renamed to `getAllGreaterOrEqualRanges`

File: src/jvmTest/kotlin/de/voize/semver4k/SemverTest.kt [hunk 4]
```diff
@@ -361,6 +361,54 @@ class SemverTest {
         Assert.assertFalse(Semver("0.1.2").isStable)
     }
 
+    @Test
+    fun findsNextIncrements() {
+        Assert.assertEquals(Semver("1.0.1"), Semver("1.0.0").nextIncrement())
+        Assert.assertEquals(Semver("1.0.0-beta.0"), Semver("1.0.0-beta").nextIncrement())
+        Assert.assertEquals(Semver("1.0.0-beta.5"), Semver("1.0.0-beta.4").nextIncrement())
+        Assert.assertEquals(Semver("1.0.0-beta.5+sha899-d8g79f87"), Semver("1.0.0-beta.4+sha899-d8g79f87").nextIncrement())
```
Comment (Legion2) [inline src/jvmTest/kotlin/de/voize/semver4k/SemverTest.kt:370]: can you also add a test for pre 1.0.0 version

**Review by Legion2 — CHANGES_REQUESTED** [review]

File: src/jvmTest/kotlin/de/voize/semver4k/SemverTest.kt (hunk 4 above)
Comment (Legion2) [reply src/jvmTest/kotlin/de/voize/semver4k/SemverTest.kt:370]: 0.1.0 for example

Comment (onigoetz) [reply src/jvmTest/kotlin/de/voize/semver4k/SemverTest.kt:370]: should be good now

**Review by Legion2 — APPROVED** [review]
Thanks for your contribution

**Comment by onigoetz** [comment]
you're welcome :)

**Pull request merged**